### PR TITLE
Move String Concat Optimization to an IR Phase

### PR DESCRIPTION
**Pull request by jdconrad**
This change removes the string concatenation optimization from the semantic phase and moves it more appropriately to an ir phase.

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultStringConcatenationOptimizationPhase.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+         int i = 0;
+
+        while (i < irStringConcatenationNode.getArgumentNodes().size()) {
+            ExpressionNode irArgumentNode = irStringConcatenationNode.getArgumentNodes().get(i);
+
+            if (irArgumentNode instanceof StringConcatenationNode) {
+                irStringConcatenationNode.getArgumentNodes().remove(i);
+                irStringConcatenationNode.getArgumentNodes().addAll(i, ((StringConcatenationNode)irArgumentNode).getArgumentNodes());
+            } else {
+                ++i;
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultStringConcatenationOptimizationPhase.java:38]: weirdo

**Review by stu-elastic — APPROVED** [review]

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultStringConcatenationOptimizationPhase.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultStringConcatenationOptimizationPhase.java]: Sigh, unweirded.

**Comment by jdconrad** [comment]
@stu-elastic Thanks for the review! Will commit as soon as CI passes.

**Comment by jdconrad** [comment]
@elasticmachine run elasticsearch-ci/1

**Pull request merged**